fix(grouping): Improve storage of GroupHash last processed event index

tkaemming opened the pull request:
- This rewrites the GroupHash last processed event index to be easier to manage the growth of data over time through TTLs and deletions.
- Only the first hash used for an event is recorded in the cache, since [we only consider the first hash when unmerging](https://github.com/getsentry/sentry/blob/79641494084f3280543713324995866b6629f817/src/sentry/tasks/unmerge.py#L143-L149) and having the event details associated with hashes that when unmerged don't actually cause anything to be ungrouped can be a point of confusion. This changes the method signature of `record_last_processed_event_id` accordingly to only take a single hash, instead of multiple.

Fixes GH-6607.

tkaemming commented on src/sentry/api/serializers/models/grouphash.py:
```diff
@@ -12,19 +12,19 @@ def get_latest_events(group_hash_list):
         )
         event_by_event_id = {
             event.event_id: event
             for event in Event.objects.filter(
```
This doesn't actually use the correct index, I'm going to fix this to use the btree on `(project_id, message_id)`.

tkaemming replied:
Fixed this in 6e0a627

tkaemming replied:
Not sure what I was on about here, this is using the `event.id` primary key and not `event.event_id`…

tkaemming commented on src/sentry/models/grouphash.py:
```diff
@@ -38,36 +38,22 @@ class Meta:
-            lambda result: result.value,
-            results,
-        )
+            client.expire(key, 7776000)  # 90d
```
Magic number alert — I'm not actually sure if this is a constant value or a setting somewhere…?

getsentry-danger commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

tkaemming commented on src/sentry/models/grouphash.py:
```diff
@@ -38,36 +38,22 @@ class Meta:
     @staticmethod
-    def record_last_processed_event_id(project_id, group_hash_ids, event_id):
-        prefix = 'last-processed-event:{}'.format(project_id)
+    def record_last_processed_event_id(group_id, group_hash_id, event_id):
```
Just remembered that the group hash group is mutable which makes this a little bit more difficult. 😓 

Going to move this to a top level key instead, I guess.

tkaemming commented:
This needs a migration plan (see above) but otherwise should be good.

tkaemming commented:
Going to add a backfill script here at @mattrobenolt's request and we'll determine the way we'd like to limit the number of events to backfill for at some point before merging.

tkaemming commented on src/sentry/south_migrations/0368_backfill_grouphash_last_processed_event_data.py:
```diff
@@ -0,0 +1,1034 @@
+
+    # Flag to indicate if this migration is too risky
+    # to run online and needs to be coordinated for offline
+    is_dangerous = False
```
I'd imagine we will want to set this on one or both of these migrations?

mattrobenolt replied:
Yes please.

tkaemming replied:
Done in 313167d6d0ae7a174ac23ef7df6f934c70472ace. Let me know what you want to do about limiting the event queryset for the backfill.

tkaemming commented on src/sentry/south_migrations/0368_backfill_grouphash_last_processed_event_data.py:
```diff
@@ -0,0 +1,1034 @@
+
+        from sentry.tasks.unmerge import get_fingerprint
+
+        step = 500
```
This was chosen totally arbitrarily.

tkaemming commented on src/sentry/south_migrations/0368_backfill_grouphash_last_processed_event_data.py:
```diff
@@ -0,0 +1,1034 @@
+            # If the class method doesn't exist, we can safely skip the migration.
+            return
+
+        from sentry.tasks.unmerge import get_fingerprint
```
This is a little weird but it's important to keep these two implementations in sync, so I just imported it here.

tkaemming commented on src/sentry/south_migrations/0368_backfill_grouphash_last_processed_event_data.py:
```diff
@@ -0,0 +1,1034 @@
+        try:
+            record_last_processed_event_id = GroupHash.record_last_processed_event_id
+        except AttributeError:
+            # If the class method doesn't exist, we can safely skip the migration.
```
There's no point in doing the work here if this method doesn't actually exist.

tkaemming commented on src/sentry/south_migrations/0368_backfill_grouphash_last_processed_event_data.py:
```diff
@@ -0,0 +1,1034 @@
+                        hash=key[1],
+                    )
+                except orm.GroupHash.DoesNotExist:
+                    print('Missing group hash record for {!r}! Skipping...'.format(key))
```
This shouldn't happen (why would a `GroupHash` be missing for an event that exists?) but I'm sure it will.

tkaemming commented on src/sentry/south_migrations/0368_backfill_grouphash_last_processed_event_data.py:
```diff
@@ -0,0 +1,1034 @@
+                        grouphash.id,
+                        event.event_id,
+                    )
+                recorded.add(key)
```
This could end up being a lot of memory. I'm not sure if its beneficial to convert `event.event_id` to an integer instead of a string?

tkaemming commented on src/sentry/south_migrations/0368_backfill_grouphash_last_processed_event_data.py:
```diff
@@ -0,0 +1,1034 @@
+
+        step = 500
+        queryset = RangeQuerySetWrapperWithProgressBar(
+            Event.objects.all(),
```
We'll want to provide a limit here and we'll need to decide if we're going to execute this with the same limit ourselves as for on-premise deployments.

tkaemming commented on src/sentry/south_migrations/0368_backfill_grouphash_last_processed_event_data.py:
```diff
@@ -0,0 +1,1034 @@
+        from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+
+        try:
+            from sentry.models import Event, GroupHash
```
These have to be imported for `EventManager` and `GroupHash.record_last_processed_event_id` respectively but aren't used for querying.

tkaemming commented:
Rebased, removed backfill migration (we will run this as not a migration) and replaced data migration with one with the latest migration sequence value

tkaemming commented on src/sentry/south_migrations/0369_remove_old_grouphash_last_processed_event_data.py:
```diff
@@ -46,12 +46,12 @@ def backwards(self, orm):
             'homepage_url': ('django.db.models.fields.URLField', [], {'max_length': '200', 'null': 'True'}),
             'id': ('sentry.db.models.fields.bounded.BoundedBigAutoField', [], {'primary_key': 'True'}),
-            'name': ('django.db.models.fields.CharField', [], {'default': "'Vocal Prawn'", 'max_length': '64', 'blank': 'True'}),
+            'name': ('django.db.models.fields.CharField', [], {'default': "'Blessed Chimp'", 'max_length': '64', 'blank': 'True'}),
```
love south

tkaemming commented:
Backfill script to be run independently: https://gist.github.com/tkaemming/a3e4aa696225a2d283eafcded8cf60ee
